File: py_system/processing/pipeline.py
"""Multimodal RAG Pipeline - Text Processing"""
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from langchain_core.documents import Document
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma, FAISS
from langchain_community.retrievers import BM25Retriever
import logging
import os

logger = logging.getLogger(__name__)

@dataclass
class ProcessedChunk:
    content: str
    source: str
    chunk_index: int
    metadata: Dict[str, Any] = field(default_factory=dict)

class TextProcessor:
    def __init__(self, chunk_size=1000, chunk_overlap=200):
        self.splitter = RecursiveCharacterTextSplitter(
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
            separators=["\n\n", "\n", ". ", "? ", "! ", " "],
            length_function=len,
        )
    
    def process_text(self, text: str, source: str) -> List[ProcessedChunk]:
        chunks = self.splitter.split_text(text)
        return [
            ProcessedChunk(
                content=chunk,
                source=source,
                chunk_index=i,
                metadata={"source": source, "chunk_index": i, "type": "text"}
            )
            for i, chunk in enumerate(chunks)
        ]
    
    def process_documents(self, documents: List[Document]) -> List[Document]:
        return self.splitter.split_documents(documents)

class EmbeddingManager:
    def __init__(self, model_name="sentence-transformers/all-MiniLM-L6-v2"):
        self.model_name = model_name
        self.embeddings = None
    
    def initialize(self):
        logger.info(f"Loading embedding model: {self.model_name}")
        self.embeddings = HuggingFaceEmbeddings(
            model_name=self.model_name,
            model_kwargs={"device": "cpu"},
            encode_kwargs={"normalize_embeddings": True}
        )
        logger.info("Embedding model loaded successfully")
        return self.embeddings
    
    def get_embeddings(self):
        if self.embeddings is None:
            self.initialize()
        return self.embeddings
    
    def embed_query(self, query: str):
        return self.get_embeddings().embed_query(query)
    
    def embed_documents(self, documents: List[str]):
        return self.get_embeddings().embed_documents(documents)

# Vector Store Manager
from pathlib import Path

class VectorStoreManager:
    def __init__(self, embeddings, persist_directory="data/vector_store", provider="chromadb"):
        self.embeddings = embeddings
        self.persist_directory = persist_directory
        self.provider = provider
        self.vector_store = None
        self._load_or_create_store()
    
    def _load_or_create_store(self):
        persist_dir = self.persist_directory
        if Path(persist_dir).exists() and any(Path(persist_dir).iterdir()):
            logger.info(f"Loading existing vector store from {persist_dir}")
            self._load_store()
        else:
            logger.info("Creating new vector store")
            self._create_store()
    
    def _create_store(self):
        if self.provider == "chromadb":
            self.vector_store = Chroma(
                collection_name="multimodal_rag",
                embedding_function=self.embeddings,
                persist_directory=self.persist_directory
            )
        elif self.provider == "faiss":
            self.vector_store = FAISS.from_texts(["init"], self.embeddings)
    
    def _load_store(self):
        if self.provider == "chromadb":
            self.vector_store = Chroma(
                collection_name="multimodal_rag",
                embedding_function=self.embeddings,
                persist_directory=self.persist_directory
            )
        elif self.provider == "faiss":
            self.vector_store = FAISS.load_local(
                self.persist_directory, 
                self.embeddings,
                allow_dangerous_deserialization=True
            )
    
    def add_documents(self, documents):
        self.vector_store.add_documents(documents)
        if self.provider == "chromadb":
            self.vector_store.persist()
        elif self.provider == "faiss":
            self.vector_store.save_local(self.persist_directory)
        logger.info(f"Added {len(documents)} documents")
    
    def as_retriever(self, search_type="similarity", k=5):
        return self.vector_store.as_retriever(
            search_type=search_type,
            search_kwargs={"k": k}
        )
    
    def similarity_search(self, query, k=5):
        return self.vector_store.similarity_search(query, k=k)
    
    def similarity_search_with_score(self, query, k=5):
        return self.vector_store.similarity_search_with_score(query, k=k)

# Hybrid Retriever
class HybridRetriever:
    def __init__(self, vector_store, documents, semantic_weight=0.7, keyword_weight=0.3):
        self.vector_store = vector_store
        self.documents = documents
        self.semantic_weight = semantic_weight
        self.keyword_weight = keyword_weight
        self._setup_retrievers()
    
    def _setup_retrievers(self):
        self.semantic_retriever = self.vector_store.as_retriever(
            search_type="similarity", search_kwargs={"k": 5}
        )
        self.bm25_retriever = BM25Retriever.from_documents(self.documents, k=5)
    
    def get_relevant_documents(self, query):
        semantic_results = self.semantic_retriever.get_relevant_documents(query)
        keyword_results = self.bm25_retriever.get_relevant_documents(query)
        
        seen = set()
        combined = []
        for doc in semantic_results:
            key = doc.page_content[:50]
            if key not in seen:
                seen.add(key)
                combined.append((doc, self.semantic_weight))
        
        for doc in keyword_results:
            key = doc.page_content[:50]
            if key not in seen:
                seen.add(key)
                combined.append((doc, self.keyword_weight))
        
        combined.sort(key=lambda x: x[1], reverse=True)
        return [doc for doc, _ in combined[:5]]

# Main Pipeline Class
class MultimodalRAGPipeline:
    def __init__(self):
        self.text_processor = TextProcessor()
        self.embedding_manager = EmbeddingManager()
        self.vector_store_manager = None
        self.documents = []
    
    def initialize(self):
        embeddings = self.embedding_manager.initialize()
        self.vector_store_manager = VectorStoreManager(embeddings)
        logger.info("RAG Pipeline initialized")
    
    def ingest_text(self, text: str, source: str = "unknown") -> int:
        chunks = self.text_processor.process_text(text, source)
        documents = [
            Document(page_content=chunk.content, metadata=chunk.metadata)
            for chunk in chunks
        ]
        self.vector_store_manager.add_documents(documents)
        self.documents.extend(documents)
        logger.info(f"Ingested {len(chunks)} chunks from {source}")
        return len(chunks)
    
    def get_retriever(self, hybrid=True):
        if hybrid:
            return HybridRetriever(
                self.vector_store_manager.vector_store,
                self.documents
            )
        return self.vector_store_manager.as_retriever()
    
    def similarity_search(self, query, k=5):
        return self.vector_store_manager.similarity_search(query, k=k)

File: src/processing/pipeline.js
/**
 * Multimodal RAG Pipeline - Text Processing
 * Handles text chunking, embeddings, and vector storage
 */

import fs from 'fs';
import { v4 as uuidv4 } from 'uuid';
import { Logger } from '../utils/index.js';

const logger = new Logger({ name: 'RAGPipeline', level: process.env.RAG_LOG_LEVEL || process.env.LOG_LEVEL || 'info' });

/**
 * @typedef {Object} ProcessedChunk
 * @property {string} content - Chunk content
 * @property {string} source - Source identifier
 * @property {number} chunk_index - Chunk index
 * @property {Object} metadata - Additional metadata
 */

class TextProcessor {
    /**
     * @param {Object} options - Options for text processing
     * @param {number} options.chunkSize - Size of text chunks
     * @param {number} options.chunkOverlap - Overlap between chunks
     */
    constructor(options = {}) {
        this.chunkSize = options.chunkSize || 1000;
        this.chunkOverlap = options.chunkOverlap || 200;
        this.separators = ['\n\n', '\n', '. ', '? ', '! ', ' '];
    }

    /**
     * Process text and split into chunks
     * @param {string} text - Text to process
     * @param {string} source - Source identifier
     * @returns {Array<ProcessedChunk>} Array of processed chunks
     */
    processText(text, source) {
        const chunks = this._splitText(text);
        return chunks.map((chunk, i) => ({
            content: chunk,
            source: source,
            chunk_index: i,
            metadata: { source, chunk_index: i, type: 'text' }
        }));
    }

    /**
     * Split text using recursive character splitting
     * @param {string} text - Text to split
     * @returns {Array<string>} Array of text chunks
     */
    _splitText(text) {
        if (!text || text.length === 0) return [];

        const chunks = [];
        let startIndex = 0;
        const textLength = text.length;

        while (startIndex < textLength) {
            let endIndex = startIndex + this.chunkSize;

            if (endIndex >= textLength) {
                chunks.push(text.substring(startIndex));
                break;
            }

            // Try to split at a separator
            let splitFound = false;
            for (const separator of this.separators) {
                const separatorIndex = text.indexOf(separator, startIndex + this.chunkSize - 200);
                if (separatorIndex !== -1 && separatorIndex < startIndex + this.chunkSize) {
                    endIndex = separatorIndex + separator.length;
                    splitFound = true;
                    break;
                }
            }

            if (!splitFound) {
                // Hard split at chunk size
                endIndex = startIndex + this.chunkSize;
            }

            chunks.push(text.substring(startIndex, endIndex).trim());
            startIndex = endIndex - this.chunkOverlap;

            if (startIndex < 0) startIndex = 0;
        }

        return chunks.filter(chunk => chunk.length > 0);
    }

    /**
     * Process documents (compatible with LangChain style)
     * @param {Array} documents - Array of document objects
     * @returns {Array} Split documents
     */
    processDocuments(documents) {
        const results = [];
        for (const doc of documents) {
            const chunks = this.processText(doc.page_content || doc.content || doc.text, doc.source || 'unknown');
            results.push(...chunks);
        }
        return results;
    }
}

class EmbeddingManager {
    /**
     * @param {Object} options - Options for embeddings
     * @param {string} options.modelName - Model name for embeddings
     */
    constructor(options = {}) {
        this.modelName = options.modelName || 'sentence-transformers/all-MiniLM-L6-v2';
        this.embeddings = null;
        this._transformers = null;
        this._pipeline = null;
    }

    /**
     * Initialize the embedding model
     */
    async initialize() {
        logger.info('Loading embedding model', { model: this.modelName });

        try {
            const { pipeline, env } = await import('@xenova/transformers');

            // Disable local model files to use remote
            env.allowLocalModels = false;
            env.useBrowserCache = true;

            this._pipeline = await pipeline('feature-extraction', this.modelName);
            logger.info('Embedding model loaded successfully');

            this.embeddings = {
                embedQuery: async (query) => {
                    const output = await this._pipeline(query, { pooling: 'mean', normalize: true });
                    return Array.from(output.data);
                },
                embedDocuments: async (documents) => {
                    const embeddings = [];
                    for (const doc of documents) {
                        const output = await this._pipeline(doc, { pooling: 'mean', normalize: true });
                        embeddings.push(Array.from(output.data));
                    }
                    return embeddings;
                }
            };

            return this.embeddings;
        } catch (e) {
            logger.error('Failed to load embedding model', { error: e.message });
            // Return a mock embedding function as fallback
            this.embeddings = {
                embedQuery: async (query) => {
                    // Simple hash-based embedding as fallback
                    const hash = this._simpleHash(query);
                    return new Array(384).fill(0).map((_, i) => Math.sin(hash + i) * 0.1);
                },
                embedDocuments: async (documents) => {
                    return Promise.all(documents.map(doc => this.embeddings.embedQuery(doc)));
                }
            };
            return this.embeddings;
        }
    }

    /**
     * Simple hash function for mock embeddings
     * @param {string} str - String to hash
     * @returns {number} Hash value
     */
    _simpleHash(str) {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            const char = str.charCodeAt(i);
            hash = ((hash << 5) - hash) + char;
            hash = hash & hash;
        }
        return hash;
    }

    /**
     * Get embeddings instance
     * @returns {Object} Embeddings object
     */
    getEmbeddings() {
        return this.embeddings;
    }

    /**
     * Embed a single query
     * @param {string} query - Query string
     * @returns {Promise<Array>} Embedding vector
     */
    async embedQuery(query) {
        if (!this.embeddings) {
            await this.initialize();
        }
        return await this.embeddings.embedQuery(query);
    }

    /**
     * Embed multiple documents
     * @param {Array<string>} documents - Array of document strings
     * @returns {Promise<Array<Array>>} Array of embedding vectors
     */
    async embedDocuments(documents) {
        if (!this.embeddings) {
            await this.initialize();
        }
        return await this.embeddings.embedDocuments(documents);
    }
}

class VectorStoreManager {
    /**
     * @param {Object} embeddings - Embeddings instance
     * @param {Object} options - Options for vector store
     */
    constructor(embeddings, options = {}) {
        this.embeddings = embeddings;
        this.persistDirectory = options.persistDirectory || './data/vector_store';
        this.provider = options.provider || 'memory';
        this.vectorStore = null;
        this.documents = [];
        this.vectors = [];
        this._loadOrCreateStore();
    }

    /**
     * Load or create vector store
     */
    _loadOrCreateStore() {
        if (fs.existsSync(this.persistDirectory)) {
            logger.info('Loading existing vector store', { path: this.persistDirectory });
            this._loadStore();
        } else {
            logger.info('Creating new vector store');
            this._createStore();
        }
    }

    /**
     * Create new vector store
     */
    _createStore() {
        this.documents = [];
        this.vectors = [];
    }

    /**
     * Load existing vector store
     */
    _loadStore() {
        try {
            const dataPath = `${this.persistDirectory}/store.json`;

            if (fs.existsSync(dataPath)) {
                const data = JSON.parse(fs.readFileSync(dataPath, 'utf-8'));
                this.documents = data.documents || [];
                this.vectors = data.vectors || [];
            } else {
                this._createStore();
            }
        } catch (e) {
            logger.error('Failed to load vector store', { error: e.message });
            this._createStore();
        }
    }

    /**
     * Save vector store to disk
     */
    _saveStore() {
        try {
            if (!fs.existsSync(this.persistDirectory)) {
                fs.mkdirSync(this.persistDirectory, { recursive: true });
            }

            const dataPath = `${this.persistDirectory}/store.json`;
            fs.writeFileSync(dataPath, JSON.stringify({
                documents: this.documents,
                vectors: this.vectors
            }, null, 2));

            logger.info('Vector store saved', { path: dataPath });
        } catch (e) {
            logger.error('Failed to save vector store', { error: e.message });
        }
    }

    /**
     * Add documents to the vector store
     * @param {Array} documents - Array of documents
     */
    async addDocuments(documents) {
        if (!documents || documents.length === 0) return;

        const texts = documents.map(doc => doc.page_content || doc.content || doc.text || '');

        try {
            const embeddings = await this.embeddings.embedDocuments(texts);

            for (let i = 0; i < documents.length; i++) {
                this.documents.push({
                    ...documents[i],
                    id: documents[i].id || uuidv4()
                });
                this.vectors.push(embeddings[i]);
            }

            this._saveStore();
            logger.info('Added documents to vector store', { count: documents.length });
        } catch (e) {
            logger.error('Failed to add documents', { error: e.message });
        }
    }

    /**
     * Get retriever for similarity search
     * @param {Object} options - Retriever options
     * @returns {Object} Retriever object
     */
    asRetriever(options = {}) {
        const searchType = options.searchType || 'similarity';
        const k = options.k || 5;

        return {
            getRelevantDocuments: async (query) => {
                return await this.similaritySearch(query, k);
            }
        };
    }

    /**
     * Perform similarity search
     * @param {string} query - Query string
     * @param {number} k - Number of results
     * @returns {Array} Search results
     */
    async similaritySearch(query, k = 5) {
        if (!this.embeddings) {
            throw new Error('Embeddings not initialized');
        }

        const queryEmbedding = await this.embeddings.embedQuery(query);

        // Calculate cosine similarity
        const similarities = this.vectors.map((vector, index) => {
            const similarity = this._cosineSimilarity(queryEmbedding, vector);
            return { index, similarity };
        });

        // Sort by similarity and get top k
        similarities.sort((a, b) => b.similarity - a.similarity);
        const topK = similarities.slice(0, k);

        return topK.map(({ index }) => this.documents[index]);
    }

    /**
     * Calculate cosine similarity between two vectors
     * @param {Array} vec1 - First vector
     * @param {Array} vec2 - Second vector
     * @returns {number} Cosine similarity
     */
    _cosineSimilarity(vec1, vec2) {
        let dotProduct = 0;
        let norm1 = 0;
        let norm2 = 0;

        for (let i = 0; i < vec1.length; i++) {
            dotProduct += vec1[i] * vec2[i];
            norm1 += vec1[i] * vec1[i];
            norm2 += vec2[i] * vec2[i];
        }

        return dotProduct / (Math.sqrt(norm1) * Math.sqrt(norm2));
    }

    /**
     * Similarity search with scores
     * @param {string} query - Query string
     * @param {number} k - Number of results
     * @returns {Array} Results with scores
     */
    async similaritySearchWithScore(query, k = 5) {
        if (!this.embeddings) {
            throw new Error('Embeddings not initialized');
        }

        const queryEmbedding = await this.embeddings.embedQuery(query);

        const similarities = this.vectors.map((vector, index) => ({
            document: this.documents[index],
            score: this._cosineSimilarity(queryEmbedding, vector)
        }));

        similarities.sort((a, b) => b.score - a.score);
        return similarities.slice(0, k);
    }
}

class HybridRetriever {
    /**
     * @param {VectorStoreManager} vectorStore - Vector store manager
     * @param {Array} documents - Array of documents
     * @param {Object} options - Retriever options
     */
    constructor(vectorStore, documents, options = {}) {
        this.vectorStore = vectorStore;
        this.documents = documents;
        this.semanticWeight = options.semanticWeight || 0.7;
        this.keywordWeight = options.keywordWeight || 0.3;
        this.semanticRetriever = vectorStore.asRetriever({ searchType: 'similarity', k: 5 });
        this.bm25Retriever = null;
        this._setupRetrievers();
    }

    /**
     * Setup retrievers
     */
    _setupRetrievers() {
        // BM25-style keyword retrieval using simple term frequency
        this.bm25Retriever = {
            getRelevantDocuments: async (query) => {
                const queryTerms = query.toLowerCase().split(/\s+/);
                const scores = this.documents.map(doc => {
                    const content = (doc.page_content || doc.content || doc.text || '').toLowerCase();
                    let score = 0;
                    for (const term of queryTerms) {
                        if (content.includes(term)) {
                            score += 1;
                        }
                    }
                    return score;
                });

                const scored = this.documents.map((doc, i) => ({ doc, score: scores[i] }));
                scored.sort((a, b) => b.score - a.score);

                return scored.slice(0, 5).filter(s => s.score > 0).map(s => s.doc);
            }
        };
    }

    /**
     * Get relevant documents
     * @param {string} query - Query string
     * @returns {Array} Relevant documents
     */
    async getRelevantDocuments(query) {
        const semanticResults = await this.semanticRetriever.getRelevantDocuments(query);
        const keywordResults = await this.bm25Retriever.getRelevantDocuments(query);

        const seen = new Set();
        const combined = [];

        for (const doc of semanticResults) {
            const key = (doc.page_content || doc.content || doc.text || '').substring(0, 50);
            if (!seen.has(key)) {
                seen.add(key);
                combined.push({ doc, weight: this.semanticWeight });
            }
        }

        for (const doc of keywordResults) {
            const key = (doc.page_content || doc.content || doc.text || '').substring(0, 50);
            if (!seen.has(key)) {
                seen.add(key);
                combined.push({ doc, weight: this.keywordWeight });
            }
        }

        combined.sort((a, b) => b.weight - a.weight);
        return combined.slice(0, 5).map(({ doc }) => doc);
    }
}

class MultimodalRAGPipeline {
    /**
     * Initialize the RAG pipeline
     */
    constructor() {
        this.textProcessor = new TextProcessor();
        this.embeddingManager = new EmbeddingManager();
        this.vectorStoreManager = null;
        this.documents = [];
    }

    /**
     * Initialize the pipeline
     */
    async initialize() {
        const embeddings = await this.embeddingManager.initialize();
        this.vectorStoreManager = new VectorStoreManager(embeddings);
        logger.info('RAG Pipeline initialized');
    }

    /**
     * Ingest text into the pipeline
     * @param {string} text - Text to ingest
     * @param {string} source - Source identifier
     * @returns {Promise<number>} Number of chunks created
     */
    async ingestText(text, source = 'unknown') {
        const chunks = this.textProcessor.processText(text, source);

        const documents = chunks.map(chunk => ({
            page_content: chunk.content,
            metadata: chunk.metadata
        }));

        await this.vectorStoreManager.addDocuments(documents);
        this.documents.push(...documents);

        logger.info('Ingested chunks', { count: chunks.length, source });
        return chunks.length;
    }

    /**
     * Get retriever for querying
     * @param {boolean} hybrid - Whether to use hybrid retrieval
     * @returns {Object} Retriever object
     */
    getRetriever(hybrid = true) {
        if (hybrid) {
            return new HybridRetriever(
                this.vectorStoreManager,
                this.documents
            );
        }
        return this.vectorStoreManager.asRetriever();
    }

    /**
     * Perform similarity search
     * @param {string} query - Query string
     * @param {number} k - Number of results
     * @returns {Promise<Array>} Search results
     */
    async similaritySearch(query, k = 5) {
        return await this.vectorStoreManager.similaritySearch(query, k);
    }
}

export { TextProcessor, EmbeddingManager, VectorStoreManager, HybridRetriever, MultimodalRAGPipeline };
export default {
    TextProcessor,
    EmbeddingManager,
    VectorStoreManager,
    HybridRetriever,
    MultimodalRAGPipeline
};
